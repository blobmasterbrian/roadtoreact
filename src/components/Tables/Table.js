// @flow strict
import "./Table.css";
import React from "react";
import { Button } from "../Buttons";

import type { Element, Node } from "react";

type TableProps = {
  children?: Node,
  list: Array<Entry>,
  onDismiss: (number) => void,
};

export type Entry = {
  author: string,
  num_comments: number,
  objectID: number,
  points: number,
  title: string,
  url: string,
};

export function Table({
  children,
  list,
  onDismiss,
}: TableProps): Element<"div"> {
  const matchesSearch: (string) => (Entry) => boolean = (searchTerm) => {
    return (entry: Entry): boolean => {
      return !entry.title
        ? false
        : entry.title.toLowerCase().includes(searchTerm.toLowerCase());
    };
  };

  return (
    <div className="table">
      {list.map((entry: Entry): Element<"div"> => (
        <div key={entry.objectID} className="table-row">
          <span style={largeColumn}>
            <a href={entry.url}>{entry.title}</a>
          </span>
          <span style={midColumn}> {entry.author} </span>
          <span style={smallColumn}> {entry.num_comments} </span>
          <span style={smallColumn}> {entry.points} </span>
          <span style={smallColumn}>
            <Button
              onClick={(): void => onDismiss(entry.objectID)}
              className="button-inline"
            >
              {children}
            </Button>
          </span>
        </div>
      ))}
    </div>
  );
}

const largeColumn: Object = {
  width: "40%",
};

const midColumn: Object = {
  width: "30%",
};

const smallColumn: Object = {
  width: "10%",
};
