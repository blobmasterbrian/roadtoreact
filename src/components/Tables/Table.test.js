// @flow strict
import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { Table } from "./Table";

import type { Entry } from "./Table";

describe("Table", () => {
  const list: Array<Entry> = [
    {
      author: "brian",
      num_comments: 207,
      objectID: 80743,
      points: 55,
      title: "react",
      url: "https://reactjs.org/",
    },
    {
      author: "blobmaster",
      num_comments: 702,
      objectID: 43317,
      points: 36,
      title: "redux",
      url: "https://redux.js.org/",
    },
  ];

  test("renders without crashing", () => {
    const { getAllByText }: Object = render(
      <Table onDismiss={() => {}} list={list}>
        Dismiss
      </Table>
    );
    const textElements: Array<HTMLElement> = getAllByText(/Dismiss/i);
    textElements.map((textElement: HTMLElement) => {
      expect(textElement).toBeInTheDocument();
    });
  });

  test("has a valid snapshot", () => {
    const component: Object = renderer.create(
      <Table onDismiss={() => {}} list={list}>
        Dismiss
      </Table>
    );
    const tree: Object = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
