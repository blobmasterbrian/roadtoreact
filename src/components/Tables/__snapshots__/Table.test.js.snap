// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://reactjs.org/"
      >
        react
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
       
      brian
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      207
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      55
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://redux.js.org/"
      >
        redux
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
       
      blobmaster
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      702
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
       
      36
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
